### ISSUE TITLE
preprocessor error messages not helpful

### ISSUE THREAD

**Issue by Bertk** [SonarOpenCommunity/sonar-cxx#838]
I recognized some useless messages like
```
ERROR: Mismatch: expected ',' got: ')'
ERROR: Mismatch: expected ',' got: ')'
ERROR: Mismatch: expected ',' got: ')'
```
The error message should also refer or print the source code line (sonar-cxx/cxx-squid/src/main/java/org/sonar/cxx/preprocessor/CxxPreprocessor.java, line 761):
```java
  private List<Token> match(List<Token> tokens, String str) throws MismatchException {
    if (!tokens.get(0).getValue().equals(str)) {
      LOG.error("Mismatch: expected '" + str + "' got: '" + tokens.get(0).getValue() + "'");
      throw new MismatchException("Mismatch: expected '" + str + "' got: '"
        + tokens.get(0).getValue() + "'");
    }
    return tokens.subList(1, tokens.size());
  }
```

**Comment by guwirth** [comment]
Agree line and col would help. Maybe we should check all messages in preprocessor and add this information.

**Comment by jmecosta** [comment]
That's my doing, however the former was even worse a stack trace that made
no sense. Agree that it's worth to try to add more info there

**Comment by Bertk** [comment]
I did an analysis and this message is floods the log file - it was written 361769 times (file total lines 408648) :worried:
I think this message is repeated most of the time for the same reason and this shall be avoided e.g. record this error message once until the reason or line changes.

**Comment by jmecosta** [comment]
You are using boost, I'm our case it was because of a macro in there

**Comment by jmecosta** [comment]
by the way, in your case before the log would be GBs, no? that was the reason for me to change this.

**Comment by jmecosta** [comment]
```
[03:41:18] :		 [Step 1/1] 03:41:18.298 ERROR - Undefined functionlike macro 'BOOST_PP_CONFIG_FLAGS' assuming 0
[03:41:18] :		 [Step 1/1] 03:41:18.298 ERROR - Mismatch: expected ',' got: ')'
[03:41:18] :		 [Step 1/1] 03:41:18.299 ERROR - Mismatch: expected ',' got: ')'
```

**Comment by Bertk** [comment]
Yes, this error logging will create huge files which cannot be loaded by an text editor :unamused: 
I used a sample from msdn [variadic_macros.cpp](https://msdn.microsoft.com/en-us/library/ms177415.aspx) for the squid unit tests and got this output:
```
com.sonar.sslr.api.RecognitionException: Parse error at line 10 column 9:

  -->  int main() {
   11:     if ( ! ( 0 ) ) { printf ( "here %s %s %s" , "are" , "some" , "varargs1(1)\n" ) ; };
   12:     if ( ! ( 1 ) )

        at org.sonar.sslr.internal.vm.Machine.parse(Machine.java:74)
        at com.sonar.sslr.impl.Parser.parse(Parser.java:87)
        at com.sonar.sslr.impl.Parser.parse(Parser.java:72)
        at org.sonar.cxx.parser.CxxParserTest.testParsingOnDiverseSourceFiles(CxxParserTest.java:55)
```
I guess the preprocessor grammar is not correct and the exception _com.sonar.sslr.api.RecognitionException_ creates a lot of noise. 
VS output for the sample looks like this and compiles without problems.
```
here are some varargs1(1)
here are some varargs2(4)
here are some varargs3(5)
hello, world
error

```

**Comment by jmecosta** [comment]
Suppose its not a bad idea to cache the parsing errors, and just print the the unique parsing error. Looked to me that that it was just repeating the same parsing error over and over?

**Comment by Bertk** [comment]
@jmecosta Yes, this will reduce the size of log files dramatically.
@guwirth The preprocessor support for the Microsoft VC specific handling of trailing commas is not working. Was this already supported?
```
The C Standard specifies that at least one argument must be passed to the ellipsis, 
to ensure that the macro does not resolve to an expression with a trailing comma. 
The Visual C++ implementation will suppress a trailing comma if no arguments 
are passed to the ellipsis.
```

**Comment by Bertk** [comment]
Sorry, I have to correct my last comment.
The Microsoft specific handling works but not this:
```
#define MACRO(s, ...) printf(s, __VA_ARGS__)

int main() {
    MACRO("hello, world\n");
}
```
The error looks like:
```
C:\GitHub\sonar-cxx>java -jar sslr-cxx-toolkit\target\sslr-cxx-toolkit-1.0-SNAPSHOT.jar
08:22:19.287 [main] INFO  o.s.c.toolkit.CxxConfigurationModel - The property 'sonar.sourceEncoding' is not set, using the default value 'UTF-8'.
08:22:26.021 [AWT-EventQueue-0] ERROR CxxPreprocessor - Mismatch: expected ',' got: ')'

```
and sslr-cxx-toolkit shows this exception:
```
com.sonar.sslr.api.RecognitionException: Parse error at line 3 column 9:

  -->  int main() {
    4:     printf ( "hello, world\n" , );
    5: }EOF

	at org.sonar.sslr.internal.vm.Machine.parse(Machine.java:74)
	at com.sonar.sslr.impl.Parser.parse(Parser.java:87)
	at com.sonar.sslr.impl.Parser.parse(Parser.java:81)
	at org.sonar.sslr.internal.toolkit.SourceCodeModel.setSourceCode(SourceCodeModel.java:62)
	at org.sonar.sslr.internal.toolkit.ToolkitPresenter.onSourceCodeParseButtonClick(ToolkitPresenter.java:124)
	at org.sonar.sslr.internal.toolkit.ToolkitViewImpl$5.actionPerformed(ToolkitViewImpl.java:203)
	at javax.swing.AbstractButton.fireActionPerformed(Unknown Source)
	at javax.swing.AbstractButton$Handler.actionPerformed(Unknown Source)
	at javax.swing.DefaultButtonModel.fireActionPerformed(Unknown Source)
	at javax.swing.DefaultButtonModel.setPressed(Unknown Source)
	at javax.swing.plaf.basic.BasicButtonListener.mouseReleased(Unknown Source)
	at java.awt.Component.processMouseEvent(Unknown Source)
	at javax.swing.JComponent.processMouseEvent(Unknown Source)
	at java.awt.Component.processEvent(Unknown Source)
	at java.awt.Container.processEvent(Unknown Source)
	at java.awt.Component.dispatchEventImpl(Unknown Source)
	at java.awt.Container.dispatchEventImpl(Unknown Source)
	at java.awt.Component.dispatchEvent(Unknown Source)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Unknown Source)
	at java.awt.LightweightDispatcher.processMouseEvent(Unknown Source)
	at java.awt.LightweightDispatcher.dispatchEvent(Unknown Source)
	at java.awt.Container.dispatchEventImpl(Unknown Source)
	at java.awt.Window.dispatchEventImpl(Unknown Source)
	at java.awt.Component.dispatchEvent(Unknown Source)
	at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
	at java.awt.EventQueue.access$500(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue.dispatchEvent(Unknown Source)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.run(Unknown Source)
```

**Comment by jmecosta** [comment]
Worth checking with previous released versions of toolkit to see if this is
a regression or not. I don't keep logs older than 2 week for the size
reason I cannot confirm. But I do remember having always big logs, but
cannot say this is the reason

**Comment by guwirth** [comment]
@Bert not sure if this was working? Please create an error issue with the sample I will have a look.

**Comment by Bertk** [comment]
Actual the case below is supported and the trailing "," is removed in   preprocessor  ```replaceParams(List<Token> body, List<Token> parameters, List<Token> arguments)``` 
```
#define EMPTY
#define MACRO(s, ...) printf(s, __VA_ARGS__)

int main() {
    MACRO("error\n", EMPTY);  // would cause error C2059, except VC++ 
                              // suppresses the trailing comma
}
```

**Comment by Bertk** [comment]
@jmecosta Please check #846. The file name and line number still needs to be added.

**Comment by Bertk** [comment]
This issue is solved with PR #846
